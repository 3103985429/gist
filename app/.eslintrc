{
  // extend root arrays
  "extends": "../.eslintrc.js",
  "parserOptions": {
    "ecmaVersion": 6,
  },
  "env": {
    "browser": false,
    "node": true
  }
}
