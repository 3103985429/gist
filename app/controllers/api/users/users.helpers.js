const async = require('async');

const User = require('../../../models/users');
const Team = require('../../../models/teams');
const mailer = require('../../../libs/utils/nodemailer');
const sample_dataset = require('../sample_dataset');

const isEnterprise = process.env.NODE_ENV === 'enterprise';

/**
 * if team is enterprise, get team object
 * @param {object} req - the req
 * @param {function} callback
 */
const getEnterpriseTeam = (req, callback) => {
    if (isEnterprise) {
        Team.findOne({ subdomain: process.env.SUBDOMAIN }, (err, team) => {
            callback(err, team, req);
        });
    } else {
        callback(null, null, req);
    }
};

/**
 * create a new team if there is no enterprise team
 * @param {object} enterpriseTeam - found enterprise team or null
 * @param {object} req - request
 * @param {function} callback
 */
const newTeam = (enterpriseTeam, req, callback) => {
    const { _id: userId, _team: team } = req.body;

    if (!enterpriseTeam) {
        team.admin = [userId];
        team.superTeam = (team.subdomain === 'schema' || isEnterprise);
        team.isEnterprise = isEnterprise;

        Team.create(team, (err, createdTeam) => {
            if (err) {
                return callback(err);
            } else if (!createdTeam) {
                return callback(new Error('Could not create team'));
            } else {
                if (createdTeam.title !== 'sampleTeam' && (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'staging')) {
                    return sample_dataset.delegateDatasetDuplicationTasks(userId, createdTeam, err => {
                        if (err) {
                            return callback(err);
                        }

                        callback(null, req, createdTeam);
                    });
                }

                callback(null, req, createdTeam);
            }
        });
    } else {
        callback(null, req, enterpriseTeam);
    }
};

/**
 * find and update the user object
 * @param {object} req - request
 * @param {object} team
 * @param {function} callback
 */
const updateUser = (req, team, callback) => {
    const { firstName, lastName, password, _id: userId } = req.body;

    User.findById(userId, (err, user) => {
        if (err || !user) {
            return callback(err || new Error(`Could not find user with id: ${userId}`));
        }

        if (req.isAnonymous && user.signedUp) {
            return callback(new Error('User already registered'));
        } else {
            user.signedUp = true;
        }

        if (user.provider === 'local' && password && (!user.hash || !user.salt)) {
            user.setPassword(password);
        }
        user.firstName = firstName;
        user.lastName = lastName;
        user._team = [team._id];
        user.defaultLoginTeam = team._id;
        callback(null, user, team);
    });
};

/**
 * find user and check if he has already signed in before
 * @param {object} req - request
 * @param {function} callback
 */
const checkUser = (req, callback) => {
    const { _id: userId } = req.body;

    User.findById(userId, (err, user) => {
        if (err || !user) {
            return callback(err || new Error(`Could not find user with id: ${userId}`));
        }

        if (user.signedUp) {
            return callback(new Error('User already registered'));
        }

        return callback(null);
    });
};

/**
 * Set activated to true.
 * Users creating their own team will need to click the email verification link to activate.
 * Invited users have already arrived at the signup page via the email link.
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const activateUser = (user, team, callback) => {
    user.activated = true;
    callback(null, user, team);
};

/**
 * save the updated user
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const saveUpdatedUser = (user, team, callback) => {
    user.save((err, updatedUser) => {
        callback(err, updatedUser, team);
    });
};

/**
 * if not enterprise, call method to send email notifying user that their team was created
 * if user is activated (aka invited user) - do nothing
 * if user is not activated (aka creating team) - call method to send an activation email link
 * @param {object} user
 * @param {object} team
 * @param {function} callback
 */
const sendNotificationEmail = (user, team, callback) => {
    if (!isEnterprise) {
        team.notifyNewTeamCreation();
    }

    if (user.activated) {
        callback(null, user);
    } else {
        mailer.sendActivationEmail(user, err => {
            callback(err, user);
        });
    }
};

/**
 * new user signup process
 */
module.exports.newUser = (req, callback) => {
    async.waterfall([
        async.apply(checkUser, req),
        async.apply(getEnterpriseTeam, req),
        newTeam,
        updateUser,
        saveUpdatedUser,
        sendNotificationEmail,
    ], (err, user) => {
        callback(err, user);
    });
};

/**
 * invited user signup process
 */
module.exports.userInvited = (req, callback) => {
    const { _team: team } = req.body;

    async.waterfall([
        async.apply(checkUser, req),
        async.apply(updateUser, req, team),
        activateUser,
        saveUpdatedUser,
    ], (err, user) => {
        callback(err, user);
    });
};
