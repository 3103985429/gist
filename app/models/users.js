const mongoose = require('mongoose');
const crypto = require('crypto');
const winston = require('winston');
const { isEmpty } = require('lodash');

const { Schema } = mongoose;

const userSchema = new Schema({
    email: {
        type: String,
        unique: true,
        required: true,
    },
    firstName: String,
    lastName: String,
    profileImageUrl: String,
    provider: String,
    hash: String,
    salt: String,
    loginAttempts: { type: Number, required: true, default: 0 },
    lockUntil: { type: Number, default: 1 },
    activated: {
        type: Boolean,
        default: false,
    },
    sampleImported: {
        type: Boolean,
        default: false,
    },
    _team: [{ type: Schema.Types.ObjectId, ref: 'Team' }],
    active: {
        type: Boolean,
        default: true,
    },
    inviteLink: String,
    // denotes those users with non-schema emails as superAdmins
    superUser: Boolean,

    invited: Object, //temp object to store invited new user
    canCreateNewViz: Array,
    canCreateNewArticle: Array,
    canCreateNewSite: Array,
    _editors: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _viewers: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    _articleEditors: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _articleViewers: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    _siteEditors: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    _siteViewers: [{ type: Schema.Types.ObjectId, ref: 'Website' }],
    defaultLoginTeam: { type: Schema.Types.ObjectId, ref: 'Team' },
    inviteToken: { type: Schema.Types.ObjectId, ref: 'Token' },
    confirmation_code_hash: Object,
    signedUp: {
        type: Boolean,
        default: false,
    },
    inviter: String,
    role: String,
    smartsheetToken: Object, // Smartsheet token object to hold smartsheet token connection information
    pipedriveToken: Object, // Pipedrive token object
    datadotworldToken: Object, // Data.world token object
    salesforceToken: Object, // Salesforce token object
}, { timestamps: true });

const deepPopulate = require('mongoose-deep-populate')(mongoose);
userSchema.plugin(deepPopulate, { whitelist: ['defaultLoginTeam.datasourceDescriptions', '_team.datasourceDescriptions'] });

userSchema.methods.setPassword = function (password) {
    this.salt = crypto.randomBytes(16).toString('hex');
    this.hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
};

userSchema.methods.validPassword = function (password) {
    const hash = crypto.pbkdf2Sync(password, this.salt, 1000, 64, 'sha512').toString('hex');
    return this.hash === hash;
};

userSchema.methods.isSuperAdmin = function () {
    const re = new RegExp('\\+superuser@schemadesign.com$');
    return re.test(this.email) || this.superUser === true;
};

userSchema.methods.isAdmin = function (team) {
    if (!team) {
        winston.warn('isAdmin was called without team param');
        return false;
    }

    if (isEmpty(team.admin)) {
        winston.warn('isAdmin was called on team without admin set');
        return false;
    }

    const userId = this._id.toString();

    return team.admin.some(id => id.toString() === userId);
};

userSchema.methods.isTeamMember = function (team) {
    if (!team) {
        winston.warn('isTeamMember was called without team param');
        return false;
    }
    return this._teams.some(team => team.equals(team._id));
};

userSchema.methods.isLocked = function () {
    // check for a future lockUntil timestamp
    return this.lockUntil && this.lockUntil > Date.now();
};

userSchema.methods.checkLoginAttempts = function () {
    // Add properties to prevent brute force login if user doesn't have them
    if (!this.lockUntil) {
        this.lockUntil = 1;
    }
    if (!this.loginAttempts) {
        this.loginAttempts = 0;
    }
};

userSchema.methods.lockoutUser = function () {
    this.loginAttempts += 1;
    if (this.loginAttempts > 5) { // Lock user out after 6 failed login attempts
        this.lockUntil = Date.now() + (24 * 60 * 60 * 1000); // Lock user out for 24 hours
    }
};

userSchema.methods.resetLock = function () {
    this.loginAttempts = 0;
    this.lockUntil = 1;
};

module.exports = mongoose.model('User', userSchema);
