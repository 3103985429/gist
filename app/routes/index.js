const { defaultTo } = require('lodash');
const helmet = require('helmet');
const express = require('express');

const { errorLogger } = require('../utils/winston');
const { checkCors, API_ALLOWED } = require('./permissions');
const { version } = require('../../package.json');

const PROTOCOL = process.env.USE_SSL === 'true' ? 'https' : 'http';
const HOST = defaultTo(process.env.HOST, 'localhost:9080');
const ROOT_DOMAIN = `${PROTOCOL}://app.${HOST}`;
const API_VERSION = 'v1';

const _mountRoutes_monitoring = function (app) {
    app.get('/_ah/health', function (req, res) {
        res.set('Content-Type', 'text/plain');
        res.status(200).send('ok');
    });
};

const _mountRoutes_errorHandling = function (app) {
    // Add the error logger after all middleware and routes so that
    // it can log errors from the whole application. Any custom error
    // handlers should go after this.
    errorLogger(app);

    // eslint-disable-next-line no-unused-vars
    app.use(function (err, req, res, next) {
        const status = err.status || 404;
        const [title, message] = (status === 404) ?
            ['Oops!', 'The page you are looking for doesn\'t exist or has been moved.'] :
            ['Uh oh.', 'Something went wrong\u2026'];
        const error = process.env.NODE_ENV === 'development' ? err.toString() : `Error: ${status}`;

        res.status(status).render('error', {
            env: process.env,
            title,
            message,
            err: error,
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });

    app.use(function (req, res) {
        res.status(404);
        res.render('error', {
            env: process.env,
            title: 'Oops!',
            message: 'The page you are looking for doesn\'t exist or has been moved.',
            err: 'Error: 404',
            has_google_analytics: !!process.env.GOOGLE_ANALYTICS_ID,
        });
    });
};

const _mountRoutes_endPoints = function (app) {
    app.all('*', function ({ subdomains, url }, res, next) {
        const redirectToRootDomainPatterns = [
            /^\/dashboard/,
            /^\/login/,
            /^\/signup/,
            /^\/auth/,
            /^\/product/,
            /^\/news/,
            /^\/pricing/,
            /^\/plans/,
        ];
        const isEnterprise = process.env.NODE_ENV === 'enterprise';
        const isRootDomain = subdomains.length === 1 && ['app', 'www'].includes(subdomains[0]);

        if (!isEnterprise && !isRootDomain && redirectToRootDomainPatterns.some(pattern => pattern.test(url))) {
            return res.redirect(ROOT_DOMAIN + url);
        }

        next();
    });

    app.use('/robots.txt', express.static(__dirname + '/../../robots.txt'));

    app.use('/', require('./homepage'));

    app.use('/s', require('./shared_pages'));
    app.use('/presentation', require('./stories'));
    app.use(`/json-api/${API_VERSION}`, checkCors(API_ALLOWED), function (req, res, next) {
        req.apiVersion = API_VERSION;
        next();
    }, require('./json-api/router'));
    app.use('/auth', helmet.frameguard(), checkCors(), require('./auth'));
    app.use('/login', helmet.frameguard(), checkCors(), function (req, res) {
        res.redirect('/auth/login');
    });
    app.use('/signup', helmet.frameguard(), checkCors(), require('./signup'));
    app.use('/reset', helmet.frameguard(), checkCors(), function (req, res) {
        // TODO: Perform an jwt auth here instead of in account.js / or better yet auth in both!
        res.render('auth/password', {
            env: process.env,
        });
    });

    app.use('/dashboard', helmet.frameguard(), checkCors(), require('./dashboard'));

    app.use('/api', helmet.frameguard(), require('./api'));

    app.use('/account', checkCors(), require('./account'));
    app.use('/edit-visualization', require('./external/editVisualization'));
    app.use('/create-visualization', require('./external/createVisualization'));
    app.use('/visualization-list', require('./external/visualizationList'));

    app.use('/', require('./articles'));
    app.use('/', checkCors(API_ALLOWED), require('./views').router);
};

module.exports.MountRoutes = function (app) {
    app.get('/env', function (req, res) {
        const obj = {
            node_env: process.env.NODE_ENV,
            host: HOST,
            s3Bucket: process.env.DO_S3_BUCKET,
            s3Domain: process.env.DO_S3_ENDPOINT,
            subdomain: process.env.SUBDOMAIN,
            authProtocol: process.env.AUTH_PROTOCOL,
            hasSubteams: process.env.SUBTEAMS,
            git: {
                branch: process.env.BRANCH_NAME,
                commit: process.env.COMMIT_HASH,
            },
            version,
            insight: process.env.INSIGHT,
            hide_business_links: process.env.HIDE_BUSINESS_LINKS,
            hasSmartsheetInfo: !!process.env.SMARTSHEET_CLIENT_ID,
            hasPipedriveInfo: !!process.env.PIPEDRIVE_CLIENT_ID,
            hasDatadotworldInfo: !!process.env.DATA_DOT_WORLD_CLIENT_ID,
            hasSalesforceInfo: !!process.env.SALESFORCE_CONSUMER_KEY
        };

        return res.json(obj);
    });

    _mountRoutes_monitoring(app);
    _mountRoutes_endPoints(app);
    _mountRoutes_errorHandling(app);
};
