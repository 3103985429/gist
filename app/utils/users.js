const Users = require('../models/users');

module.exports = {
    getUser,
};

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getUser(id) {
    return await Users
        .findById(id)
        .lean()
        .exec();
}
