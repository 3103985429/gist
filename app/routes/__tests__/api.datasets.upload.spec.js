import request from 'supertest';
import path from 'path';

import app from '../../../app';
import { hydrateDb, getId } from '../../../internals/testing/backend/utils';
import { user1 } from '../../../internals/testing/backend/fixtures/users';
import {
    pokemon1,
    freshViz,
    newVizJson
} from '../../../internals/testing/backend/fixtures/datasets';

const prepareRawDatasets = (files, ext) => files.reduce((acc, fileName) => Object.assign(acc, {
    [fileName]: {
        name: `${fileName}.${ext}`,
        path: path.resolve(__dirname, `../../../internals/testing/backend/fixtures/rawDatasets/${fileName}.${ext}`),
    },
}), {});

const user1Id = getId(user1);
const pokemon1Id = getId(pokemon1);
const freshVizId = getId(freshViz);
const newVizJsonId = getId(newVizJson);
const rawDatasetsCsv = prepareRawDatasets([
    'pokemonGO',
    'childPokemon',
    'datasetNoColumnNames',
    'wrongCSVFormat',
    'multiLine',
], 'csv');
const rawDatasetsJson = prepareRawDatasets([
    'allDataTypes',
    'nestedArray'
], 'json');


describe('Datasets API', () => {
    let appRequest;

    beforeEach(async () => {
        await hydrateDb();
        appRequest = request(app);
    });

    describe('Upload', () => {
        it(`should upload dataset: ${rawDatasetsCsv.pokemonGO.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', freshVizId)
                .attach('file', rawDatasetsCsv.pokemonGO.path, rawDatasetsCsv.pokemonGO.name);

            expect(response).toMatchSnapshot();
        });

        it(`should upload child dataset: ${rawDatasetsCsv.childPokemon.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', pokemon1Id)
                .field('child', 'true')
                .attach('file', rawDatasetsCsv.childPokemon.path, rawDatasetsCsv.childPokemon.name);

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                _id: expect.any(String),
            });
        });

        it(`should upload multiline dataset: ${rawDatasetsCsv.multiLine.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', pokemon1Id)
                .attach('file', rawDatasetsCsv.multiLine.path, rawDatasetsCsv.multiLine.name);

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                _id: expect.any(String),
            });
        });

        it(`should upload a replacement file: ${rawDatasetsCsv.pokemonGO.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', user1Id)
                .field('id', pokemon1Id)
                .field('replacement', 'true')
                .attach('file', rawDatasetsCsv.pokemonGO.path, rawDatasetsCsv.pokemonGO.name);

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                _id: expect.any(String),
            });
        });

        it('should fail when id of datasource not found', async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', '5a42c26e303770207ce8f83f');

            expect(response).toMatchSnapshot();
        });

        it('should fail when file extension differs from file format', async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', pokemon1Id)
                .attach('file', rawDatasetsCsv.pokemonGO.path, 'pokemonGO.json');

            expect(response.statusCode).toBe(500);
            expect(response).toMatchSnapshot();
        });

        it('should fail when upload a replacement file when file extension differs from file format', async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', pokemon1Id)
                .attach('file', rawDatasetsCsv.pokemonGO.path, 'pokemonGO.json');

            expect(response).toMatchSnapshot();
        });

        it(`should fail when upload a csv file with wrong format: ${rawDatasetsCsv.wrongCSVFormat.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', pokemon1Id)
                .attach('file', rawDatasetsCsv.wrongCSVFormat.path, rawDatasetsCsv.wrongCSVFormat.name);

            expect(response).toMatchSnapshot();
        });

        it('should fail when upload the replacement with file of different columns numbers', async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', user1Id)
                .field('id', pokemon1Id)
                .field('replacement', 'true')
                .attach('file', rawDatasetsCsv.datasetNoColumnNames.path, rawDatasetsCsv.datasetNoColumnNames.name);

            expect(response).toMatchSnapshot();
        });

        it(`should upload valid JSON dataset: ${rawDatasetsJson.allDataTypes.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', newVizJsonId)
                .attach('file', rawDatasetsJson.allDataTypes.path, rawDatasetsJson.allDataTypes.name);

            expect(response).toMatchSnapshot();
        });

        it('should fail on wrong JSON', async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', newVizJsonId)
                .attach('file', Buffer.from('trash}', 'utf8'), 'sampleJson.json');

            expect(response).toMatchSnapshot();
        });

        it(`should upload valid nested JSON dataset: ${rawDatasetsJson.nestedArray.name}`, async () => {
            const response = await appRequest
                .post('/api/dataset/upload')
                .field('user', JSON.stringify(user1))
                .field('id', newVizJsonId)
                .attach('file', rawDatasetsJson.allDataTypes.path, rawDatasetsJson.allDataTypes.name);

            expect(response).toMatchSnapshot();
        });
    });
});
