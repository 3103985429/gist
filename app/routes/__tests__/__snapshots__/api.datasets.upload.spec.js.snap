// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Datasets API Upload should fail on wrong JSON 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Invalid JSON (Unexpected \\"a\\" at position 2 in state STOP)",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should fail when file extension differs from file format 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Invalid JSON (Unexpected \\"P\\" at position 0 in state STOP)",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should fail when id of datasource not found 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Could not find datasource description.",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should fail when upload a csv file with wrong format: wrongCSVFormat.csv 1`] = `
Object {
  "body": Object {
    "error": "Wrong number of columns, expected at least 2, found 1",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should fail when upload a replacement file when file extension differs from file format 1`] = `
Object {
  "body": Object {
    "error": "Non-public error: Invalid JSON (Unexpected \\"P\\" at position 0 in state STOP)",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should fail when upload the replacement with file of different columns numbers 1`] = `
Object {
  "body": Object {
    "error": "Datasources are not compatible. Can't have fewer columns than previous datasource.",
  },
  "statusCode": 500,
}
`;

exports[`Datasets API Upload should upload a replacement file: pokemonGO.csv 1`] = `
Object {
  "_id": Any<String>,
  "columns": Array [
    Object {
      "data_type": "Text",
      "name": "Image URL",
      "operation": "ToString",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Number",
      "name": "Max HP",
      "operation": "ToInteger",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Number",
      "name": "Max CP",
      "operation": "ToInteger",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Text",
      "name": "Type 2",
      "operation": "ToString",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Text",
      "name": "Type 1",
      "operation": "ToString",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Text",
      "name": "Name",
      "operation": "ToString",
      "sourceName": "pokemonGO.csv",
    },
    Object {
      "data_type": "Number",
      "name": "Pokemon No_",
      "operation": "ToInteger",
      "sourceName": "pokemonGO.csv",
    },
  ],
  "fe_excludeFields": Object {
    "Image URL": false,
    "Max CP": false,
    "Max HP": false,
    "Name": false,
    "Pokemon No_": false,
    "Type 1": false,
    "Type 2": false,
  },
  "fe_excludeFieldsObjDetail": Object {
    "Image URL": false,
    "Max CP": false,
    "Max HP": false,
    "Name": false,
    "Pokemon No_": false,
    "Type 1": false,
    "Type 2": false,
  },
  "fileName": "pokemonGO.csv",
  "format": "CSV",
  "jobId": 0,
  "missingSamples": Array [],
  "overrideColumns": Array [],
  "raw_rowObjects_coercionScheme": Object {
    "Image URL": Object {
      "operation": "ToString",
    },
    "Max CP": Object {
      "operation": "ToInteger",
    },
    "Max HP": Object {
      "operation": "ToInteger",
    },
    "Name": Object {
      "operation": "ToString",
    },
    "Pokemon No_": Object {
      "operation": "ToInteger",
    },
    "Type 1": Object {
      "operation": "ToString",
    },
    "Type 2": Object {
      "operation": "ToString",
    },
  },
  "replaced_id": "5a42c26e303770207ce8f83b",
  "replacement": true,
  "uid": "pokemon-draft",
  "user": Object {
    "__v": 0,
    "_articleEditors": Array [],
    "_articleViewers": Array [],
    "_editors": Array [],
    "_id": "5a42bf07d3bc0b20648290e0",
    "_siteEditors": Array [],
    "_siteViewers": Array [],
    "_team": Array [
      "5a42bf30d3bc0b20648290e1",
    ],
    "_viewers": Array [],
    "activated": true,
    "active": true,
    "canCreateNewArticle": Array [],
    "canCreateNewSite": Array [],
    "canCreateNewViz": Array [],
    "createdAt": "2017-12-26T21:28:39.946Z",
    "defaultLoginTeam": "5a42bf30d3bc0b20648290e1",
    "email": "user@schemadesign.com",
    "firstName": "Maitland",
    "hash": "dc449e96c930ca46653ec867c448719fab39dd85c4b91c1643d2faad140f24e711da13133e34ee4da0d3a5a2ca154edd4bfaeb9cf80cb9af57bce4d7ea1b94ce",
    "lastName": "McConnell",
    "lockUntil": 1,
    "loginAttempts": 0,
    "provider": "local",
    "salt": "test",
    "sampleImported": false,
    "signedUp": false,
    "updatedAt": "2017-12-26T21:36:31.420Z",
  },
}
`;

exports[`Datasets API Upload should upload child dataset: childPokemon.csv 1`] = `
Object {
  "_id": Any<String>,
  "columns": Array [
    Object {
      "data_type": "Number",
      "name": "Pokemon No",
      "operation": "ToInteger",
      "sample": 152,
      "sourceName": "childPokemon.csv",
      "sourceType": "spreadsheet",
    },
    Object {
      "data_type": "Text",
      "name": "Name",
      "operation": "ToString",
      "sample": "Maitland",
      "sourceName": "childPokemon.csv",
      "sourceType": "spreadsheet",
    },
  ],
  "fe_excludeFields": Object {
    "Name": false,
    "Pokemon No": false,
  },
  "fe_excludeFieldsObjDetail": Object {
    "Name": false,
    "Pokemon No": false,
  },
  "fileName": "childPokemon.csv",
  "format": "CSV",
  "jobId": 0,
  "missingSamples": Array [],
  "overrideColumns": Array [],
  "raw_rowObjects_coercionScheme": Object {
    "Name": Object {
      "operation": "ToString",
    },
    "Pokemon No": Object {
      "operation": "ToInteger",
    },
  },
  "replacement": false,
  "uid": "pokemon",
}
`;

exports[`Datasets API Upload should upload dataset: pokemonGO.csv 1`] = `
Object {
  "body": Object {
    "columns": Array [
      Object {
        "data_type": "Number",
        "name": "Pokemon No_",
        "operation": "ToInteger",
        "sample": 1,
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "Name",
        "operation": "ToString",
        "sample": "Bulbasaur",
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "Type 1",
        "operation": "ToString",
        "sample": "Grass",
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "Type 2",
        "operation": "ToString",
        "sample": "Poison",
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Number",
        "name": "Max CP",
        "operation": "ToInteger",
        "sample": 1079,
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Number",
        "name": "Max HP",
        "operation": "ToInteger",
        "sample": 83,
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "Image URL",
        "operation": "ToString",
        "sample": "http://cdn.bulbagarden.net/upload/thumb/2/21/001Bulbasaur.png/250px-001Bulbasaur.png",
        "sourceName": "pokemonGO.csv",
        "sourceType": "spreadsheet",
      },
    ],
    "fe_excludeFields": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
    },
    "fe_excludeFieldsObjDetail": Object {
      "Image URL": false,
      "Max CP": false,
      "Max HP": false,
      "Name": false,
      "Pokemon No_": false,
      "Type 1": false,
      "Type 2": false,
    },
    "fileName": "pokemonGO.csv",
    "format": "CSV",
    "jobId": 0,
    "missingSamples": Array [],
    "overrideColumns": Array [],
    "raw_rowObjects_coercionScheme": Object {
      "Image URL": Object {
        "operation": "ToString",
      },
      "Max CP": Object {
        "operation": "ToInteger",
      },
      "Max HP": Object {
        "operation": "ToInteger",
      },
      "Name": Object {
        "operation": "ToString",
      },
      "Pokemon No_": Object {
        "operation": "ToInteger",
      },
      "Type 1": Object {
        "operation": "ToString",
      },
      "Type 2": Object {
        "operation": "ToString",
      },
    },
    "replacement": false,
    "uid": "newviztest",
  },
  "statusCode": 200,
}
`;

exports[`Datasets API Upload should upload multiline dataset: multiLine.csv 1`] = `
Object {
  "_id": Any<String>,
  "columns": Array [
    Object {
      "data_type": "Number",
      "name": "line",
      "operation": "ToInteger",
      "sample": 1,
      "sourceName": "multiLine.csv",
      "sourceType": "spreadsheet",
    },
    Object {
      "data_type": "Text",
      "name": "text",
      "operation": "ToString",
      "sample": "this line has more than one line of text",
      "sourceName": "multiLine.csv",
      "sourceType": "spreadsheet",
    },
  ],
  "fe_excludeFields": Object {
    "line": false,
    "text": false,
  },
  "fe_excludeFieldsObjDetail": Object {
    "line": false,
    "text": false,
  },
  "fileName": "multiLine.csv",
  "format": "CSV",
  "jobId": 0,
  "missingSamples": Array [],
  "overrideColumns": Array [],
  "raw_rowObjects_coercionScheme": Object {
    "line": Object {
      "operation": "ToInteger",
    },
    "text": Object {
      "operation": "ToString",
    },
  },
  "replacement": false,
  "uid": "pokemon",
}
`;

exports[`Datasets API Upload should upload valid JSON dataset: allDataTypes.json 1`] = `
Object {
  "body": Object {
    "columns": Array [
      Object {
        "data_type": "Date",
        "input_format": "YYYY-MM-DD",
        "name": "Date",
        "operation": "ToDate",
        "output_format": "YYYY-MM-DD",
        "sample": "2018-01-01T00:00:00.000Z",
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Number",
        "name": "Number",
        "operation": "ToInteger",
        "sample": 1,
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "String",
        "operation": "ToString",
        "sample": "IAmString1",
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
    ],
    "fe_excludeFields": Object {
      "Date": false,
      "Number": false,
      "String": false,
    },
    "fe_excludeFieldsObjDetail": Object {
      "Date": false,
      "Number": false,
      "String": false,
    },
    "fileName": "allDataTypes.json",
    "format": "JSON",
    "jobId": 0,
    "missingSamples": Array [],
    "overrideColumns": Array [],
    "raw_rowObjects_coercionScheme": Object {
      "Date": Object {
        "format": "YYYY-MM-DD",
        "operation": "ToDate",
        "outputFormat": "YYYY-MM-DD",
      },
      "Number": Object {
        "operation": "ToInteger",
      },
      "String": Object {
        "operation": "ToString",
      },
    },
    "replacement": false,
    "uid": "newviz-json",
  },
  "statusCode": 200,
}
`;

exports[`Datasets API Upload should upload valid nested JSON dataset: nestedArray.json 1`] = `
Object {
  "body": Object {
    "columns": Array [
      Object {
        "data_type": "Date",
        "input_format": "YYYY-MM-DD",
        "name": "Date",
        "operation": "ToDate",
        "output_format": "YYYY-MM-DD",
        "sample": "2018-01-01T00:00:00.000Z",
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Number",
        "name": "Number",
        "operation": "ToInteger",
        "sample": 1,
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
      Object {
        "data_type": "Text",
        "name": "String",
        "operation": "ToString",
        "sample": "IAmString1",
        "sourceName": "allDataTypes.json",
        "sourceType": "spreadsheet",
      },
    ],
    "fe_excludeFields": Object {
      "Date": false,
      "Number": false,
      "String": false,
    },
    "fe_excludeFieldsObjDetail": Object {
      "Date": false,
      "Number": false,
      "String": false,
    },
    "fileName": "allDataTypes.json",
    "format": "JSON",
    "jobId": 0,
    "missingSamples": Array [],
    "overrideColumns": Array [],
    "raw_rowObjects_coercionScheme": Object {
      "Date": Object {
        "format": "YYYY-MM-DD",
        "operation": "ToDate",
        "outputFormat": "YYYY-MM-DD",
      },
      "Number": Object {
        "operation": "ToInteger",
      },
      "String": Object {
        "operation": "ToString",
      },
    },
    "replacement": false,
    "uid": "newviz-json",
  },
  "statusCode": 200,
}
`;
